src/main/java/com/uid2/admin/audit/QLDBAuditMiddleware.java

```diff
@@ -0,0 +1,134 @@
+package com.uid2.admin.audit;
+
+import com.amazon.ion.IonStruct;
+import com.amazon.ion.IonSystem;
+import com.amazon.ion.IonValue;
+import com.amazon.ion.system.IonSystemBuilder;
+import com.uid2.admin.auth.AdminUserProvider;
+import com.uid2.shared.auth.IAuthorizable;
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.web.RoutingContext;
+import software.amazon.awssdk.services.qldbsession.QldbSessionClient;
+import software.amazon.qldb.QldbDriver;
+import software.amazon.qldb.RetryPolicy;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class QLDBAuditMiddleware implements AuditMiddleware{
+
+    public static final String LEDGER_NAME = "audit-logs";
+    public static final String TABLE_NAME = "logs";
+    private static final IonSystem ionSys = IonSystemBuilder.standard().build();
+    private static final QldbDriver qldbDriver = QldbDriver.builder()
+            .ledger(LEDGER_NAME)
+            .transactionRetryPolicy(RetryPolicy.builder().maxRetries(3).build())
+            .sessionClientBuilder(QldbSessionClient.builder())
+            .build();
+    private static final Logger logger = LoggerFactory.getLogger(QLDBAuditMiddleware.class);
+
+    protected QLDBAuditMiddleware(){
+
+    }
+
+    @Override
+    public Handler<RoutingContext> handle(AuditHandler<RoutingContext> handler) {
+        InnerAuditHandler auditHandler = new InnerAuditHandler(handler, this);
+        return auditHandler::handle;
+    }
+
+    @Override
+    public void writeLog(AuditModel model) {
+        if(model == null){
+            return;
+        }
+        // write to QLDB; update this to UPDATE in one command instead of two
+        qldbDriver.execute(txn -> {
```

Comment:
can this throw exception

Reply:
If credentials fail; I'll put this in a try/catch